ef-validation/README.md

```diff
@@ -1,3 +1,279 @@
+        Validator.ValidateObject(product, validationContext, validateAllProperties: true);
+        if (await _dbContext.Products.AnyAsync(b => b.Id == product.Id && 
+        b.ProductNr != product.ProductNr 
+        && b.Name == product.Name 
+        && b.Price == product.Price))
+        {
+            throw new ValidationException(new ValidationResult($"Produkt mit den Namen {product.Name} existiert bereits", new List<string> {"Isbn"}), 
+            null, new List<string> {"Isbn"});
+        }
+    }
+}
+```
+```c#
+public async Task<int> SaveChangesAsync()
+{
+    var entities = _dbContext.ChangeTracker.Entries()
+    .Where(entity => entity.State == EntityState.Added || entity.State == EntityState.Modified)
+    .Select(e => e.Entity);
+    foreach(var entity in entities)
+    {
+        await ValidateEntity(entity);
+    }
+    return await _dbContext.SaveChangesAsync();
+}
+```
+
+## WPF
+
+* Um die WPF Validierung nutzen zu können benötigt man im zugehörigen ViewModel das Interface INotifyDataErrorInfo.
+
+* HasErrors (Property):
+    * Sind Fehler am ViewModel vorhanden?
+* ErrorsChanged (EventHandler):
+    * Bei Änderungen der Validität des ViewModels
+* GetErrors (Methode):
+    * Liefert pro Binding-Property die (!)
+    Fehlermeldung
+
+* Wichtig ist zudem auch den UpdateSourceTrigger auf PropertyChanged zu setzen um nach jedem Zeichen die Eingabe zu validieren.
+
+![](UpdateSourceTriggerPropertyChanged.png)
+
+
+### Validierungsfehler
+
+* Um Validierungsfehler beim Speichern des Contexts auszuwerten, muss die ValidationException im ViewModel abgefangen werden.
+* Sollten Fehler auftreten werden diese einem ViewModel-Property (DbError) zugewiesen, welches am Window gebunden ist.
+
+
+```c#
+try
+{
+    await uow.SaveChangesAsync();
+}
+catch (ValidationException ve)
+{
+    if (ve.Value is IEnumerable<string> properties)
+    {
+        foreach (var property in properties)
+        {
+            Errors.Add(property, new List<string> {ve.ValidationResult.ErrorMessage});
+        }
+    }
+    else
+    {
+        DbError = ve.ValidationResult.ToString();
+    }
+}
+```
+
+### ValidationAttribute
+* Diese Klasse erzwingt die Validierung basierend auf den Metadaten, die der Datentabelle zugeordnet sind. Sie kann überschrieben werden, um benutzerdefinierte Validierungsattribute zu erstellen.
+    * Beispiele: 
+        * Equals()
+        * IsValid()
+        * Validate()
+        * ...
+
+#### Beispiel Validate(): 
+* Diese wird in den Properties aufgerufen, in welchen sie benötigt wird.
+
+```c#
+class EditCustomerViewModel : BaseViewModel
+{
+        public string Lastname
+        {
+            get => _lastname;
+            set
+            {
+                _lastname = value;
+                OnPropertyChanged(nameof(Lastname));
+                Validate();
+            }
+        }
+}
+```
+
+* Die Validate-Methode der BaseViewModel wird überschrieben
+* Sie wird zur Validierung der einzelnen Properties verwendet.
```

Comment:
... verwendet und stellt eine Alternative zur Validierung per Validierungunsattribute in den Modellklassen dar!